app/invocation/cache_requests_card.tsx

```diff
@@ -113,29 +151,162 @@ export default class CacheRequestsCardComponent extends React.Component<CacheReq
+  }
+  private onChangeResponseType(event: React.ChangeEvent<HTMLSelectElement>) {
+    router.setQueryParam("response", event.target.value);
+  }
+  private onToggleGroupByAction() {
+    router.setQueryParam("groupByAction", !this.getGroupByAction());
+  }
+  private searchTimeout = 0;
+  private onChangeSearch(event: React.ChangeEvent<HTMLInputElement>) {
+    const searchText = event.target.value;
+    // Update the search box state immediately, but debounce the query update so
+    // we don't fetch on each keystroke.
+    this.setState({ searchText });
+    clearTimeout(this.searchTimeout);
+    this.searchTimeout = window.setTimeout(
+      () => router.setQueryParam("search", searchText),
+      SEARCH_DEBOUNCE_INTERVAL_MS
+    );
+  }
+
   private onClickLoadMore() {
     this.fetchResults();
   }
 
+  private renderControls() {
+    return (
+      <>
+        <div className="controls row">
+          {/* Sorting controls */}
+          <label>Sort by</label>
+          <Select value={this.getOrderBy()} onChange={this.onChangeOrderBy.bind(this)}>
+            <Option value={cache.GetCacheScoreCardRequest.OrderBy.ORDER_BY_START_TIME}>Start time</Option>
+            <Option value={cache.GetCacheScoreCardRequest.OrderBy.ORDER_BY_DURATION}>Duration</Option>
+          </Select>
+          <OutlinedButton className="icon-button" onClick={this.onToggleDescending.bind(this)}>
+            {this.getDescending() ? <SortDesc className="icon" /> : <SortAsc className="icon" />}
+          </OutlinedButton>
+          {/* Filtering controls */}
+          <div className="separator" />
+          <Select value={this.getCacheType()} onChange={this.onChangeCacheType.bind(this)}>
+            <Option value={cache.CacheType.UNKNOWN_CACHE_TYPE}>CAS and AC</Option>
+            <Option value={cache.CacheType.CAS}>CAS only</Option>
+            <Option value={cache.CacheType.AC}>AC only</Option>
+          </Select>
+          <Select value={this.getRequestType()} onChange={this.onChangeRequestType.bind(this)}>
+            <Option value={cache.RequestType.UNKNOWN_REQUEST_TYPE}>Reads and writes</Option>
+            <Option value={cache.RequestType.READ}>Reads only</Option>
+            <Option value={cache.RequestType.WRITE}>Writes only</Option>
+          </Select>
+          <Select value={this.getResponseType()} onChange={this.onChangeResponseType.bind(this)}>
+            <Option value={cache.ResponseType.UNKNOWN_RESPONSE_TYPE}>Any status</Option>
+            <Option value={cache.ResponseType.OK}>OK only</Option>
+            <Option value={cache.ResponseType.NOT_FOUND}>NotFound only</Option>
+            <Option value={cache.ResponseType.ERROR}>Errors only</Option>
+          </Select>
+          {/* Grouping controls */}
+          <div className="separator" />
+          <CheckboxButton checked={this.getGroupByAction()} onClick={this.onToggleGroupByAction.bind(this)}>
+            Group by action
```

Comment:
Should this be "Group by target"?

Reply:
Ah okay, this makes sense - I thought we were just grouping by the target name that we got in the metadata header.

Reply:
Yeah that seems reasonable

Reply:
I'll do this in another PR since it'll need some backend changes too.